clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -373,34 +374,108 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
+      // iterator i is expected to be within bounds [0,VALUES_PER_THREAD)
+      return adjust_by_padding(begin[i]);
+    }
+  };
+
+  template <typename Item, typename offsetFunctorTypeFW,
+            typename offsetFunctorTypeRV>
+  __dpct_inline__ void helper_exchange(Item item, T (&keys)[VALUES_PER_THREAD],
+                                       offsetFunctorTypeFW &offset_functor_fw,
+                                       offsetFunctorTypeRV &offset_functor_rv) {
+
     T *buffer = reinterpret_cast<T *>(_local_memory);
 
 #pragma unroll
-    for (int i = 0; i < VALUES_PER_THREAD; i++) {
-      int offset = ranks[i];
-      if (INSERT_PADDING)
-        offset = detail::shr_add(offset, LOG_LOCAL_MEMORY_BANKS, offset);
+    for (size_t i = 0; i < VALUES_PER_THREAD; i++) {
+      size_t offset = offset_functor_fw(item, i);
       buffer[offset] = keys[i];
     }
 
     item.barrier(sycl::access::fence_space::local_space);
 
 #pragma unroll
-    for (int i = 0; i < VALUES_PER_THREAD; i++) {
-      int offset = (item.get_local_id(0) * VALUES_PER_THREAD) + i;
-      if (INSERT_PADDING)
-        offset = detail::shr_add(offset, LOG_LOCAL_MEMORY_BANKS, offset);
+    for (size_t i = 0; i < VALUES_PER_THREAD; i++) {
+      size_t offset = offset_functor_rv(item, i);
       keys[i] = buffer[offset];
     }
   }
 
+  /// Rearrange elements from blocked order to striped order
+  template <typename Item>
+  __dpct_inline__ void blocked_to_striped(Item item,
+                                          T (&keys)[VALUES_PER_THREAD]) {
+
+    get_blocked_offset getBlockedOffset;
+    get_striped_offset getStripedOffset;
+    helper_exchange(item, keys, getBlockedOffset, getStripedOffset);
+  }
+
+  /// Rearrange elements from striped order to blocked order
+  template <typename Item>
+  __dpct_inline__ void striped_to_blocked(Item item,
+                                          T (&keys)[VALUES_PER_THREAD]) {
+
+    get_striped_offset getStripedOffset;
+    get_blocked_offset getBlockedOffset;
+    helper_exchange(item, keys, getStripedOffset, getBlockedOffset);
+  }
+
+  /// Rearrange elements from rank order to blocked order
+  template <typename Item>
+  __dpct_inline__ void scatter_to_blocked(Item item,
+                                          T (&keys)[VALUES_PER_THREAD],
+                                          int (&ranks)[VALUES_PER_THREAD]) {
+
+    get_scatter_offset<int *> getScatterOffset(ranks);
+    get_striped_offset getStripedOffset;
+    helper_exchange(item, keys, getScatterOffset, getStripedOffset);
+  }
+
+  /// Rearrange elements from scatter order to striped order
+  template <typename Item>
+  __dpct_inline__ void scatter_to_striped(Item item,
+                                          T (&keys)[VALUES_PER_THREAD],
+                                          int (&ranks)[VALUES_PER_THREAD]) {
+
+    get_scatter_offset<int *> getScatterOffset(ranks);
+    get_blocked_offset getBlockedOffset;
```

Comment:
Why are we using a blocked offset if we want to store in a striped arrangement?

Reply:
Addressed.